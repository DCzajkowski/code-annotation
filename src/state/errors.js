export const initialState = [];

export const ADD = 'ca/errors/ADD';
export const REMOVE = 'ca/errors/REMOVE';

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD:
      return [...state, action.error];
    case REMOVE:
      return [
        ...state.slice(0, action.index),
        ...state.slice(action.index + 1),
      ];
    default:
      return state;
  }
};

export const add = error => dispatch => {
  console.error(error);
  if (Array.isArray(error)) {
    error.forEach(e =>
      dispatch({
        type: ADD,
        error: e,
      })
    );
    return;
  }
  dispatch({
    type: ADD,
    error,
  });
};

export const remove = index => ({
  type: REMOVE,
  index,
});

export default reducer;
