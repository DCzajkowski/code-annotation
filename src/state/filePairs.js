import { LOCATION_CHANGED, replace } from 'redux-little-router';
import api from '../api';
import { add as addErrors } from './errors';
import { namedRoutes, makeUrl } from './routes';
import { load as expLoad } from './experiment';
import { load as featuresLoad } from './features';

/* reducer */

export const initialState = {
  loading: true,
  list: [],
  pairs: {},

  currentPairId: null,
  currentAnnotations: {},
};

export const LOAD = 'ca/filePairs/LOAD';
export const LOAD_SUCCESS = 'ca/filePairs/LOAD_SUCCESS';
export const LOAD_FILE_PAIR = 'ca/filePairs/LOAD_FILE_PAIR';
export const LOAD_ANNOTATIONS = 'ca/filePairs/LOAD_ANNOTATIONS';
export const SET_ANNOTATIONS = 'ca/filePairs/SET_ANNOTATIONS';
export const SET_FILE_PAIR = 'ca/filePairs/SET_FILE_PAIR';
export const SET_CURRENT_PAIR = 'ca/filePairs/SET_CURRENT_PAIR';

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case LOAD:
      return initialState;
    case LOAD_SUCCESS:
      return {
        ...state,
        loading: false,
        list: action.pairs,
      };
    case LOAD_FILE_PAIR:
      return {
        ...state,
        loading: true,
      };
    case SET_FILE_PAIR:
      return {
        ...state,
        loading: false,
        pairs: {
          ...state.pairs,
          [action.id]: {
            diff: action.diff,
            score: action.score,
            leftBlobId: action.leftBlobId,
            rightBlobId: action.rightBlobId,
            leftLoc: action.leftLoc,
            rightLoc: action.rightLoc,
          },
        },
      };
    case LOAD_ANNOTATIONS:
      return {
        ...state,
        currentAnnotations: {},
      };
    case SET_ANNOTATIONS:
      return {
        ...state,
        currentAnnotations: action.data,
      };
    case SET_CURRENT_PAIR:
      return {
        ...state,
        currentPairId: action.id,
      };

    default:
      return state;
  }
};

export default reducer;

/* Actions */

export const load = expId => dispatch => {
  dispatch({ type: LOAD });

  return api.getFilePairs(expId).then(pairs =>
    dispatch({
      type: LOAD_SUCCESS,
      pairs,
    })
  );
};

export const loadFilePairIfNeeded = (expId, id) => (dispatch, getState) => {
  const { filePairs } = getState();
  if (filePairs.pairs[id]) {
    return Promise.resolve();
  }

  dispatch({ type: LOAD_FILE_PAIR });

  return api.getFilePair(expId, id).then(res =>
    dispatch({
      type: SET_FILE_PAIR,
      ...res,
    })
  );
};

export const selectPair = (expId, id) => dispatch => {
  dispatch({
    type: SET_CURRENT_PAIR,
    id,
  });

  return dispatch(loadFilePairIfNeeded(expId, id)).catch(e => {
    dispatch(addErrors(e));
  });
};

export const loadAnnotations = (expId, id) => dispatch => {
  dispatch({ type: LOAD_ANNOTATIONS });
  return api
    .getFilePairAnnotations(expId, id)
    .then(res =>
      dispatch({
        type: SET_ANNOTATIONS,
        data: res,
      })
    )
    .catch(e => {
      dispatch(addErrors(e));
    });
};

/* Selectors */

export const getCurrentFilePair = state => {
  const { currentPairId, pairs } = state.filePairs;
  if (!currentPairId) {
    return null;
  }
  return pairs[currentPairId];
};

/* Routing */

export const middleware = store => next => action => {
  if (action.type !== LOCATION_CHANGED) {
    return next(action);
  }

  const result = next(action);
  const { payload } = action;
  const { experiment } = store.getState();
  let expIdParam;
  let promise;
  switch (payload.route) {
    case namedRoutes.review:
      expIdParam = +payload.params.experiment;
      return next(expLoad(expIdParam)).then(() =>
        next(load(expIdParam)).then(() => {
          const { filePairs } = store.getState();
          // FIXME(max) can be empty, but I really really want to fix it in separate PR
          // I'm sure it's not the only place where we handle empty responses incorrectly
          const pairId = filePairs.list[0].id;
          return next(
            replace(
              makeUrl('reviewPair', {
                experiment: expIdParam,
                pair: pairId,
              })
            )
          );
        })
      );
    case namedRoutes.reviewPair:
      expIdParam = +payload.params.experiment;
      promise = Promise.resolve();
      if (experiment.id !== expIdParam) {
        promise = next(expLoad(expIdParam)).then(() => next(load(expIdParam)));
      }
      return promise
        .then(() => next(selectPair(expIdParam, +payload.params.pair)))
        .then(() => next(loadAnnotations(expIdParam, +payload.params.pair)))
        .then(() => {
          const pair = getCurrentFilePair(store.getState());
          return pair && next(featuresLoad(pair.leftBlobId, pair.rightBlobId));
        });
    default:
      return result;
  }
};
