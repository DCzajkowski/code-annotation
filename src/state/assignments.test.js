import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { push } from 'redux-little-router';
import reducer, {
  ANSWER_SIMILAR,
  ANSWER_MAYBE,
  ANSWER_DIFFERENT,
  ANSWER_SKIP,
  initialState,
  LOAD,
  LOAD_SUCCESS,
  LOAD_ERROR,
  SET_ASSIGNMENTS,
  SET_CURRENT_ASSIGMENT,
  MARK_ASSIGNMENT,
  selectAssigment,
  nextAssigment,
  load,
  mark,
  markCurrent,
  getIdenticalCount,
  getSimilarCount,
  getDifferentCount,
  getSkipCount,
  getProgressPercent,
  getOverallTime,
  getAverageTime,
  getCurrentFilePair,
} from './assignments';
import { LOAD_FILE_PAIR, SET_FILE_PAIR } from './filePairs';
import { makeUrl } from './routes';
import { ADD as ERROR_ADD } from './errors';

const mockStore = configureMockStore([thunk]);

describe('assignments/reducer', () => {
  const RealDate = Date;

  function mockDate(d) {
    global.Date = class extends RealDate {
      constructor() {
        return new RealDate(d);
      }
    };
  }

  afterEach(() => {
    global.Date = RealDate;
  });

  it('LOAD', () => {
    expect(reducer(initialState, { type: LOAD })).toMatchSnapshot();
  });

  it('LOAD_SUCCESS', () => {
    expect(
      reducer(
        { ...initialState, loading: true },
        {
          type: LOAD_SUCCESS,
          index: 1,
        }
      )
    ).toMatchSnapshot();
  });

  it('LOAD_ERROR', () => {
    expect(
      reducer(
        { ...initialState, loading: true, fileLoading: true },
        { type: LOAD_ERROR, error: 'test error' }
      )
    ).toMatchSnapshot();
  });

  it('SET_ASSIGNMENTS', () => {
    expect(
      reducer(initialState, { type: SET_ASSIGNMENTS, assignments: [1, 2] })
    ).toMatchSnapshot();
  });

  it('LOAD_FILE_PAIR', () => {
    expect(reducer(initialState, { type: LOAD_FILE_PAIR })).toMatchSnapshot();
  });

  it('SET_FILE_PAIR', () => {
    expect(
      reducer(
        { ...initialState, loading: true },
        {
          type: SET_FILE_PAIR,
          id: 1,
          diff: 'test diff',
          score: 0.92,
          leftBlobId: 1,
          rightBlobId: 2,
        }
      )
    ).toMatchSnapshot();
  });

  it('SET_CURRENT_ASSIGMENT', () => {
    mockDate(1516188409733);

    expect(
      reducer(initialState, {
        type: SET_CURRENT_ASSIGMENT,
        assigment: { id: 1, name: 'test' },
      })
    ).toMatchSnapshot();
  });

  describe('MARK_ASSIGNMENT', () => {
    const currentAssigmentStartTime = new Date(1516188409733);

    it('existing assigment', () => {
      mockDate(1516188500000);
      expect(
        reducer(
          {
            ...initialState,
            list: [{ id: 1 }],
            currentAssigmentStartTime,
          },
          { type: MARK_ASSIGNMENT, id: 1, answer: 'yes' }
        )
      ).toMatchSnapshot();
    });

    it('unknown assigment', () => {
      mockDate(1516188500000);
      expect(
        reducer(
          {
            ...initialState,
            list: [{ id: 1 }],
            currentAssigmentStartTime,
          },
          { type: MARK_ASSIGNMENT, id: 2, answer: 'yes' }
        )
      ).toMatchSnapshot();
    });
  });
});

describe('assignments/actions', () => {
  const expId = 1;

  describe('selectAssigment', () => {
    const assigment = { id: 1, pairId: 1 };

    it('with file pair in cache', () => {
      const store = mockStore({
        assignments: {
          ...initialState,
          list: [assigment],
        },
        filePairs: {
          pairs: { 1: 'do not care about content here' },
        },
      });

      return store.dispatch(selectAssigment(expId, 1)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: SET_CURRENT_ASSIGMENT,
            assigment,
          },
        ]);
      });
    });

    it('success', () => {
      const store = mockStore({
        assignments: {
          ...initialState,
          list: [{ id: 1, pairId: 1 }],
        },
        filePairs: {
          pairs: {},
        },
      });

      fetch.mockResponse(
        JSON.stringify({
          data: {
            id: 1,
            diff: 'some diff',
          },
        })
      );

      return store.dispatch(selectAssigment(expId, 1)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: SET_CURRENT_ASSIGMENT,
            assigment,
          },
          {
            type: LOAD_FILE_PAIR,
          },
          {
            type: SET_FILE_PAIR,
            id: 1,
            diff: 'some diff',
          },
        ]);
      });
    });

    it('error', () => {
      const store = mockStore({
        assignments: {
          ...initialState,
          list: [{ id: 1, pairId: 1 }],
        },
        filePairs: {
          pairs: {},
        },
      });

      fetch.mockReject('some error');

      return store.dispatch(selectAssigment(expId, 1)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: SET_CURRENT_ASSIGMENT,
            assigment,
          },
          {
            type: LOAD_FILE_PAIR,
          },
          {
            type: ERROR_ADD,
            error: 'some error',
          },
          {
            type: LOAD_ERROR,
            error: ['some error'],
          },
        ]);
      });
    });
  });

  describe('nextAssigment', () => {
    it('has unanswered assignments', () => {
      const list = [
        { id: 1, answer: 'yes' },
        { id: 2, answer: null },
        { id: 3, answer: 'no' },
      ];

      const store = mockStore({
        assignments: {
          ...initialState,
          list,
        },
      });

      store.dispatch(nextAssigment(expId));
      expect(store.getActions()).toEqual([push('/exp/1/2')]);
    });

    it('no unanswered assignments', () => {
      const list = [{ id: 1, answer: 'yes' }];

      const store = mockStore({
        assignments: {
          ...initialState,
          list,
        },
      });

      store.dispatch(nextAssigment(expId));
      expect(store.getActions()).toEqual([push('/exp/1/finish')]);
    });
  });

  describe('load', () => {
    it('success', () => {
      const store = mockStore({
        assignments: initialState,
      });

      fetch.mockResponse(
        JSON.stringify({
          data: [
            {
              id: 1,
              answer: null,
            },
          ],
        })
      );
      return store.dispatch(load(expId)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: LOAD,
          },
          {
            type: SET_ASSIGNMENTS,
            assignments: [{ answer: null, id: 1 }],
          },
          {
            type: LOAD_SUCCESS,
          },
        ]);
      });
    });

    it('error', () => {
      const store = mockStore({
        assignments: initialState,
      });

      fetch.mockReject('some error');
      return store.dispatch(load(expId)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: LOAD,
          },
          {
            type: ERROR_ADD,
            error: 'some error',
          },
          {
            type: LOAD_ERROR,
            error: ['some error'],
          },
        ]);
      });
    });
  });

  describe('mark', () => {
    it('success', () => {
      const store = mockStore({
        assignments: {
          ...initialState,
          list: [{ id: 1 }],
        },
      });

      fetch.mockResponse(JSON.stringify({ data: {} }));
      return store.dispatch(mark(expId, 1, 'yes')).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: MARK_ASSIGNMENT,
            id: 1,
            answer: 'yes',
          },
          push(makeUrl('finish', { experiment: 1 })),
        ]);
      });
    });

    it('error', () => {
      const store = mockStore({
        assignments: {
          ...initialState,
          list: [{ id: 1 }],
        },
      });

      fetch.mockReject('some error');
      return store.dispatch(mark(expId, 1, 'yes')).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: MARK_ASSIGNMENT,
            id: 1,
            answer: 'yes',
          },
          {
            type: ERROR_ADD,
            error: 'some error',
          },
          {
            type: LOAD_ERROR,
            error: ['some error'],
          },
        ]);
      });
    });
  });

  it('markCurrent', () => {
    const store = mockStore({
      experiment: {
        id: expId,
      },
      assignments: {
        ...initialState,
        list: [{ id: 1 }],
        currentAssigment: { id: 1 },
      },
    });

    fetch.mockResponse(JSON.stringify({ data: {} }));
    return store.dispatch(markCurrent('yes')).then(() => {
      expect(store.getActions()[0]).toEqual({
        type: MARK_ASSIGNMENT,
        id: 1,
        answer: 'yes',
      });
    });
  });
});

describe('assignments/selectors', () => {
  const state = {
    assignments: {
      list: [
        { id: 1, answer: ANSWER_SIMILAR, duration: 2000 },
        { id: 2, answer: ANSWER_SIMILAR, duration: 1000 },
        { id: 3, answer: null },
        { id: 4, answer: ANSWER_DIFFERENT, duration: 5000 },
        { id: 5, answer: ANSWER_SKIP, duration: 1 },
        { id: 6, answer: ANSWER_SKIP, duration: 1 },
        { id: 7, answer: ANSWER_SKIP, duration: 1 },
        { id: 8, answer: ANSWER_MAYBE, duration: 1500 },
      ],
    },
  };

  it('getIdenticalCount', () => {
    expect(getIdenticalCount(state)).toEqual(2);
  });

  it('getSimilarCount', () => {
    expect(getSimilarCount(state)).toEqual(1);
  });

  it('getDifferentCount', () => {
    expect(getDifferentCount(state)).toEqual(1);
  });

  it('getSkipCount', () => {
    expect(getSkipCount(state)).toEqual(3);
  });

  it('getProgressPercent', () => {
    expect(getProgressPercent(state)).toEqual(50);
  });

  it('getProgressPercent empty', () => {
    expect(
      getProgressPercent({
        assignments: {
          list: [],
        },
      })
    ).toEqual(0);
  });

  it('getOverallTime', () => {
    expect(getOverallTime(state)).toEqual(9503);
  });

  it('getAverageTime', () => {
    expect(getAverageTime(state)).toEqual(1358);
  });

  it('getCurrentFilePair', () => {
    const pairState = {
      assignments: {
        currentAssigment: { id: 1, pairId: 1 },
      },
      filePairs: {
        pairs: { 1: 'something' },
      },
    };
    expect(getCurrentFilePair(pairState)).toEqual('something');
  });
});
