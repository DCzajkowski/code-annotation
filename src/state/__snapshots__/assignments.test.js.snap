// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignments/reducer LOAD 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": null,
  "list": Array [],
  "loading": true,
}
`;

exports[`assignments/reducer LOAD_ERROR 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": "test error",
  "fileLoading": true,
  "list": Array [],
  "loading": false,
}
`;

exports[`assignments/reducer LOAD_FILE_PAIR 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": null,
  "list": Array [],
  "loading": true,
}
`;

exports[`assignments/reducer LOAD_SUCCESS 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": null,
  "list": Array [],
  "loading": false,
}
`;

exports[`assignments/reducer MARK_ASSIGNMENT existing assigment 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": 2018-01-17T11:26:49.733Z,
  "error": null,
  "list": Array [
    Object {
      "answer": "yes",
      "duration": 90267,
      "id": 1,
    },
  ],
  "loading": true,
}
`;

exports[`assignments/reducer MARK_ASSIGNMENT unknown assigment 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": 2018-01-17T11:26:49.733Z,
  "error": null,
  "list": Array [
    Object {
      "id": 1,
    },
  ],
  "loading": true,
}
`;

exports[`assignments/reducer SET_ASSIGNMENTS 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": null,
  "list": Array [
    1,
    2,
  ],
  "loading": true,
}
`;

exports[`assignments/reducer SET_CURRENT_ASSIGMENT 1`] = `
Object {
  "currentAssigment": Object {
    "id": 1,
    "name": "test",
  },
  "currentAssigmentStartTime": 2018-01-17T11:26:49.733Z,
  "error": null,
  "list": Array [],
  "loading": true,
}
`;

exports[`assignments/reducer SET_FILE_PAIR 1`] = `
Object {
  "currentAssigment": null,
  "currentAssigmentStartTime": null,
  "error": null,
  "list": Array [],
  "loading": true,
}
`;
